### DISCUSSION TITLE
깃허브 사용 팁

### DISCUSSION THREAD

**Discussion by youngmin-gwon** [PLAIF-dev/.github#7] [Ideas]
깃허브를 편하게 사용할 수 있는 팁의 thread를 여기서 모아 보아요!

**Comment by youngmin-gwon** [comment]
### LGTM 매번 적기 너무 귀찮아요...

`LGTM👍` 같이 자주 쓰이는 메세지를 매번 적기 귀찮다고요? 템플릿으로 만들어서 사용해보세요!

1. `Settings -> Saved replies -> Add a saved reply` 로 이동
2. 자주 쓰이는 메시지 등록 후 저장하기

#### Reference
자세한 내용은 [여기](https://docs.github.com/en/get-started/writing-on-github/working-with-saved-replies/creating-a-saved-reply)서 확인할 수 있답니다.

 ---
**🫰추가 팁**: 단축키만으로 입력하기
1. <kbd>Ctrl</kbd> + <kbd>.</kbd>로 `saved reply` 목록을 띄우고, 
2. <kbd>Ctrl</kbd> + <kbd>번호</kbd>로 `자동 입력` 할 수 있답니다.

**Reply by pjs102793** [reply]
LGTM 👍

**Comment by youngmin-gwon** [comment]
### PR 승인 받고 일일이 합치는게 너무 귀찮아요...

Pull Request는 코드를 합치는 우아한 방법이지만, 귀찮을 때가 많습니다. 다음과 같은 것들요.

```markdown
1. 리뷰어에게 코드 승인을 받은 후에 다시 찾아가서 `merge`를 해야하는 것
2. merge 후에 더이상 쓰지 않는 `branch`를 삭제해야 하는 것
```

더 이상 귀찮아하지 않아도 됩니다. Pull Request를 만들 때, 한번만 설정하고 더 이상 신경쓰지 않아도 되거든요! 몇 가지 설정이면 되요.

#### How to do it
![setup-automerge](https://docs.github.com/assets/cb-153703/mw-1440/images/help/pull_requests/enable-auto-merge-drop-down.webp)
1. merge 타입을 설정한 후 `enable auto-merge` 버튼 클릭
2. 자동 병합 될 때, 남겨질 `메시지 작성`하기(자세하게 작성해주면 너무 좋아요!

#### Issue?
네? 제 창에서는 안보인다구요? 맞습니다. 특별한 설정을 하지 않으면 자동 병합할 수 없을거에요. 레포를 변경할 수 있는 팀장님들에게 이 기능을 쓸 수 있게 요청하세요. 다음은 팀장님들이 하셔야할 프로세스 입니다.

1. `Repository -> Settings -> Pull Request` 로 이동
2. `Allow auto-merge` 체크
3. (원하는 경우) `Automatically delete head branches` 체크

#### Reference
자세한 내용은 다음에서 확인해보세요.
1. [관리자: 레포 자동 병합 활성화](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-auto-merge-for-pull-requests-in-your-repository)
4. [사용자: 자동 병합 설정](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request)

**Reply by powerstrong** [reply]
@youngmin-gwon 와 좋은 기능입니다! 알아서 척척 찾아와주셔서 감사합니다!

**Comment by pjs102793** [comment]
# 내 리파지토리와 core_template 연동을 쉽게 하는 법

평소에 작업을 하다가 github에 issue나 PR을 보내려고 하는데, 다른 리파지토리랑 비교했을때 템플릿 자체가 초라했던 적이 있을 겁니다.

이는 템플릿 패치가 이뤄지고 있는 core_template가 내 리파지토리에 최신화가 안되고 있다는 것 일겁니다.
내 리파지토리에 최신 core_template을 업데이트 해봅시다!

## Remote
<img width="525" alt="스크린샷 2023-06-14 오후 8 40 35" src="https://github.com/PLAIF-dev/.github/assets/53421022/1000e984-3eb8-49f7-bcce-c70769a825d6">

위의 명령어를 통해 우선 내가 작업하고 있는 리파지토리에서 remote 정보를 가져와봅시다. 사진과는 다르게 origin에 대한 remote 정보 밖에 없을 겁니다.
그 이유는 내 리파지토리에 core_template가 remote되어 있지 않는 것을 의미합니다.

`git remote add template https://github.com/PLAIF-dev/core_template.git` 명령어를 실행하시면 위와 같이 origin과 template에 대한 remote가 저장될 것입니다.

그러면 끝인가요? 리모트된 정보를 로컬에 최신화 해야합니다.
`git fetch template`를 통해서 remote container를 로컬에 가져옵니다.
이러면 내 리파지토리에 core_template 리파지토리를 가져올 준비가 끝났습니다!

## Merge
이제 core_template을 연동했으니 내 리파지토리와 merge하는 과정을 거쳐야합니다.
여기서 중요한 점은 core_template과 내 리파지토리를 merge 할때 다른 branch를 파서 작업을 하는 게 좋습니다.
저는 주로 `upstream_merge_template`라는 branch를 판 후에 템플릿과 그 브랜치를 merge시켜서 PR을 날려서 main에 합칩니다.

`git checkout -b upstream_merge_template` 명령어를 통해 브랜치를 만들어주고 브랜치로 작업 영역을 바꿉시다.
그 브랜치에 `git merge --squash --allow-unrelated-histories template/main`명령어를 입력하여 합쳐 줍시다.
squash 옵션을 한 이유는 core_template에 있는 commit 내용을 하나로 줄이는 역할을 합니다. 무수한 commit 기록이 남겨지는 것보다 간결한게 좋겠죠?
allow-unrelated-histories 옵션을 준 이유는 git에서 서로 다른 프로젝트라고 인식을 해서 나오는 결과입니다. 하지만 core_template에는 .github에 관련된 작업만 하는 템플릿이기 때문에 큰 걱정은 안하셔도 됩니다.

## Conflict
merge를 할 때 conflict가 많이 날 수도 있고, 안 날수도 있지만 conflict가 났을때를 대비해서 작성합니다. conflict가 나면 기존에 git을 작업할 때랑 분위기가 많이 다를겁니다. conflict 해결은 [VSCode의 병합 편집기](https://itteamb.blogspot.com/2022/10/vs-codegit-vs-code-git-conflict.html)에서 진행 해도 되고 [명령줄](https://docs.github.com/ko/pull-requests/collaborating-with-pull-requests/addressing-merge-conflicts/resolving-a-merge-conflict-using-the-command-line)을 통해서 진행해도 됩니다.

Conflict를 해결 할 때는 core_template에서 최신화 된 부분과 내 리파지토리에서 template를 수정한 부분을 적절히 융합해서 나만의 최신 템플릿을 만들어가면 됩니다!

conflict를 해결하고 commit 작성을 한 후에 upstream_merge_template를 push하면 github에서 PR하는 방식으로 처리 하면 됩니다!

그 다음에는 생각 날때마다, 중요한 core_template 패치가 이뤄지었거나 등등 이럴때에 `git fetch template`에서 PR을 반복해서 패치해나가시면 됩니다.

## 요약
1. `git remote add template https://github.com/PLAIF-dev/core_template.git`
2. `git fetch template`
3. `git checkout -b upstream_merge_template`
4. `git merge --squash --allow-unrelated-histories template/main`
5.  conflict가 발생하면 각자 재량으로 conflict 해결~
6. `git push`후 PR을 통해서 의견 조율 후에 main과 merge
7. 생각 날 때마다 2 ~ 6 반복

**Reply by youngmin-gwon** [reply]
다같이 공유하는 지식으로 만들어주셔서 감사합니다! 덕분에 많은 분들이 작업물과 템플릿을 합칠 수 있게 되었네요 💯

**Reply by powerstrong** [reply]
ㅎㅎ 감사합니다 템플릿 반짝반짝하게 잘 관리하겠습니다 👍

**Comment by pjs102793** [comment]
# Slack과 GitHub 연동하기
Slack과 GitHub을 연동하면 GitHub에서 발생하는 알림들을 Slack 채널이나, 개인적인 앱 채널에 전송할 수 있습니다. 이를 통해 팀원들이 GitHub의 활동을 실시간으로 확인할 수 있게 되어 협업의 효율성을 높일 수 있습니다.

## 1. Slack에 GitHub 앱 설치하기
<img width="250" alt="스크린샷 2023-09-11 오전 11 35 30" src="https://github.com/PLAIF-dev/.github/assets/53421022/ce2ce0eb-67d1-4a59-a26e-46c392a2ea95">

Slack의 좌측 하단의 앱 추가를 클릭하고, 앱 페이지에서에서 "GitHub"을 검색합니다.

<img width="354" alt="스크린샷 2023-09-11 오전 11 36 03" src="https://github.com/PLAIF-dev/.github/assets/53421022/e57dd61c-065f-4bbe-85f4-bf4115c4d198">

이렇게 생긴 블록에 추가버튼을 클릭합니다. 저는 Github가 이미 설치되어 있어 다른걸로 예시를 들었습니다.

## 2. GitHub 앱 설정하기
<img width="648" alt="스크린샷 2023-09-11 오전 11 26 13" src="https://github.com/PLAIF-dev/.github/assets/53421022/c75b576c-23ba-4905-8d1e-364dc5504a56">

설치가 완료되면, Github 앱 채팅에 들어가서, 위와 같이 Connect Github account를 클릭하여 연동합니다. 그리고 웹페이지에 출력되는 코드를 Slack의 Enter code 버튼을 눌러서 입력합니다.

## 3. Github 리파지토리 구독
Github의 특정 리파지토리를 구독하여 그 리파지토렝 관한 알림을 받을 수 있습니다.
명령은 간단합니다.

<img width="619" alt="스크린샷 2023-09-11 오전 11 29 26" src="https://github.com/PLAIF-dev/.github/assets/53421022/c9b66e1f-0a24-4fb5-acb2-4dc3a9aed438">

저는 DLManager를 자주 사용하니 PLAIF-dev에 있는 DLManager를 구독했습니다. 구독하면 기본적으로, issues, pulls, commits, releases, deployments에 관한 모든 알림이 Slack Github 앱 채널에 올라오게 됩니다.

## 4. 연동 확인하기
GitHub에서 테스트용 이슈나 풀 리퀘스트를 생성하여 Slack 채널에 알림이 올바르게 전송되는지 확인합니다.

<img width="664" alt="스크린샷 2023-09-11 오전 11 32 07" src="https://github.com/PLAIF-dev/.github/assets/53421022/e9e97205-0205-4cca-81ad-c02e8b997d74">

이슈를 생성하고, 닫는 과정이 잘 알림이 왔습니다. 이 외에도 다른 규칙도 모두 잘 오는 것을 확인 하실 수 있을겁니다.

## 추가 팁
Slack 앱 채팅으로 개인적으로 연동 할 수 있지만, 앱을 채널에 추가하면, 채널에 있는 사람 모두가 Github 내용에 대해서 공유받을 수 있습니다. 이것도 단순히 채널에 앱을 초대하고 위의 과정을 똑같이 진행하면 됩니다. 이러면 다른 분께서 Github를 사용하실 때, 즉각적인 피드백이 가능케 하여, 더욱 편안한 협업을 진행 할 수 있을겁니다.

이해가 안되는 내용이나, 따로 궁금하신 점이 있으면 reply 부탁드립니다~

**Reply by youngmin-gwon** [reply]
## 추가 + 추가 팁

누군가 github PR 을 보내거나 댓글에 언급한 사실을 까맣게 모르고 있다가, 몇일 후 발견하여 당황한 경험이 있으실 겁니다. 간단한 `slack real-time notification` 으로 이런 일을 방지할 수 있습니다.

## How to do

1. [reminder 설정](https://github.com/settings/reminders/PLAIF-dev) 이동

2. `real-time alerts` 설정 활성화하기
<img width="1552" alt="Screenshot 2023-09-11 at 11 44 34 AM" src="https://github.com/PLAIF-dev/.github/assets/52237605/efc42071-bfbc-4522-831d-a33161e17c5d">

3. 필요한 설정 활성화 하기(ex. 누군가 나를 언급했을 때, merge conflict가 생겼을 때, etc.)  

4. 설정 저장하기

**Reply by pjs102793** [reply]
이런 적이 엄청 많았는데, 이런 기능도 있군요!

어마어마한 꿀팁 감사합니다 😄

**Reply by powerstrong** [reply]
보람찬 개발생활을 위한 꿀팁들이군요. 리스펙합니다 😇

**Reply by powerstrong** [reply]
https://github.com/integrations/slack#scheduled-reminders
매뉴얼 본 건 첨인데 이것저것 기능이 있군요 ㅋㅋ
